accepted/wasmtime-baseline-compilation.md

```diff
@@ -0,0 +1,258 @@
+mostly factored out into a separate crate `cranelift_asm`. This will require
+some care with respect to layering: in particular, definitions of
+machine-instruction types are currently done in the ISLE backends for each ISA
+within Cranelift. We can continue to use ISLE for these, but they will need to
+be moved to the separate crate.
+
+
+As a result of this initial layering, one will be able to build a `MachInst` as
+a Rust data structure and emit it manually, for example:
+
+```rust
+let add = cranelift_asm::x64::AluRmiR { op: AluRmiR::Add, … };
+let mut buf = MachBuffer::new();
+add.emit(&mut buf, …);
+```
+
+However this is still quite cumbersome. As a next step, we will develop an API
+over this that provides for procedural generation of instructions: i.e., one
+method call for each instruction. Something like:
+
+```rust
+let mut masm = cranelift_asm::x64::MacroAssembler::new(); 
+masm.add(rd, rm);
+masm.store(rd, MemArg::base_offset(ra, 64));
+``` 
+This would allow for
+fairly natural single-pass code emission. In essence, this is an implementation
+of the [MacroAssembler
+idea](https://searchfox.org/mozilla-central/rev/fa71140041c5401b80a11f099cc0cd0653295e2c/js/src/jit/MacroAssembler.h)
+from SpiderMonkey. Each architecture will have an implementation of the
+MacroAssembler API; perhaps there can be a trait that abstracts commonalities,
+but enough will be different (e.g., instruction set quirks beyond the usual
+“add/sub/and/or/not” suspects, x64 two-operand form vs aarch64 three-operand
+form, and more) that we expect there to be different `MacroAssembler` types for
+each ISA. This in turn implies different lowering code that invokes the
+`MacroAssembler` per ISA in the baseline compiler. The lowering code can perhaps
+share many helpers that are monomorphized on the “common ISA core” trait.
+
+In the above examples, we bypass the register-allocation support, i.e. the
+ability to hold virtual register operands rather than real registers, in the
+`MachInst`s. This is supported today by passing through `RealReg`s (“real
+registers”) instead. In the baseline compiler we expect register allocation to
+occur before invoking the `MacroAssembler`; i.e., when generating the
+instructions we already know which register we are using for each operand. Doing
+otherwise (emitting with vregs first and editing later) requires actually
+buffering the MachInst structs in memory, which we do not wish to do.
+
+We don’t expect to make any changes to Cranelift itself beyond the layering
+refactor to borrow its `MachInst` and `MachBuffer` implementations. In
+particular we don’t expect to use the `MacroAssembler` wrapper in Cranelift, at
+least at first, because it will be built around constructing and emitting
+instructions to machine code right away, without buffering (as in Cranelift’s
+VCode). It’s possible in the future that we may find other ways to make
+`MacroAssembler` generic and leverage it in Cranelift too, but that is beyond
```

Comment:
If `MacroAssembler` isn't used by Cranelift, I think it should not be put in `cranelift_asm`, but either in wince or a separate crate.

Reply:
It feels natural to have the MacroAssembler at this level in my opinion, given that `cranelift_asm`'s objective is to provide the functionality and building blocks for code emission from different scenarios. That said, I don't think that this is a one-way door decision, if it's proven to cause any issues, this is easily reversible.

Reply:
That's a fair concern; on the other hand, I'm confident that we can control exposing the MacroAssembler with a compile time feature, making it opt-in to consumers.